File: SatinalmaPro.Shared/Procurement/TabFilterManager.cs
using SatinalmaPro.Shared.Models;

namespace SatinalmaPro.Shared.Procurement;

/// <summary>
/// Rol ve sekmeye göre Firestore sorgu şartlarını tek merkezden üretir.
/// Android (Kotlin) ve WPF aynı kuralları kullanır.
/// </summary>
public static class TabFilterManager
{
    private static readonly ProcurementTab[] FullProcurementTabs =
    [
        ProcurementTab.PendingDraft,
        ProcurementTab.ManagementApproval,
        ProcurementTab.QuoteRequested,
        ProcurementTab.QuoteReview,
        ProcurementTab.ApprovedOrders,
        ProcurementTab.Rejected,
        ProcurementTab.HistoryReports
    ];

    public static string NormalizeRole(string? role)
    {
        var normalized = KullaniciRolleri.Normalize(role);
        return normalized switch
        {
            KullaniciRolleri.Admin => "admin",
            KullaniciRolleri.Yonetim => "yonetim",
            KullaniciRolleri.Satinalma => "satinalma",
            KullaniciRolleri.Sef => "sef",
            KullaniciRolleri.Saha => "saha",
            KullaniciRolleri.Depo => "depo",
            KullaniciRolleri.Atolye => "atolye",
            _ => normalized.ToLowerInvariant()
        };
    }

    public static bool RequiresRequesterScope(string? role)
    {
        var key = NormalizeRole(role);
        return key is "sef" or "saha";
    }

    public static IReadOnlyList<ProcurementTab> GetVisibleTabs(string? role)
    {
        return NormalizeRole(role) switch
        {
            "depo" =>
            [
                ProcurementTab.ApprovedOrders,
                ProcurementTab.HistoryReports,
                ProcurementTab.StockMovements
            ],
            "atolye" => [ProcurementTab.StockStatus],
            "admin" or "yonetim" or "satinalma" => FullProcurementTabs,
            "sef" or "saha" => FullProcurementTabs,
            _ => FullProcurementTabs
        };
    }

    public static bool IsTabVisible(string? role, ProcurementTab tab) =>
        GetVisibleTabs(role).Contains(tab);

    public static FirestoreFilterSpec? GetQuerySpec(
        ProcurementTab tab,
        string? role,
        string? currentUid)
    {
        if (!IsTabVisible(role, tab))
            return null;

        var requesterUid = RequiresRequesterScope(role) ? currentUid : null;
        if (RequiresRequesterScope(role) && string.IsNullOrWhiteSpace(requesterUid))
            return null;

        return tab switch
        {
            ProcurementTab.PendingDraft => new FirestoreFilterSpec
            {
                Collection = "procurement_requests",
                StatusIn = [ProcurementStatus.Draft, ProcurementStatus.Submitted],
                RequesterUidEquals = requesterUid,
                OrderBy =
                [
                    new FirestoreOrderBy { Field = "updatedAtUtc", Descending = true }
                ]
            },
            ProcurementTab.ManagementApproval => new FirestoreFilterSpec
            {
                Collection = "procurement_requests",
                StatusIn = [ProcurementStatus.Submitted],
                RequesterUidEquals = requesterUid,
                UrgentFirst = true,
                OrderBy =
                [
                    new FirestoreOrderBy { Field = "priority", Descending = true },
                    new FirestoreOrderBy { Field = "updatedAtUtc", Descending = true }
                ]
            },
            ProcurementTab.QuoteRequested => new FirestoreFilterSpec
            {
                Collection = "procurement_requests",
                StatusIn =
                [
                    ProcurementStatus.QuoteRequested,
                    ProcurementStatus.QuoteEntry,
                    ProcurementStatus.Comparison
                ],
                RequesterUidEquals = requesterUid,
                OrderBy = [new FirestoreOrderBy { Field = "updatedAtUtc", Descending = true }]
            },
            ProcurementTab.QuoteReview => new FirestoreFilterSpec
            {
                Collection = "procurement_requests",
                StatusIn = [ProcurementStatus.ManagementQuoteReview],
                RequesterUidEquals = requesterUid,
                OrderBy = [new FirestoreOrderBy { Field = "updatedAtUtc", Descending = true }]
            },
            ProcurementTab.ApprovedOrders => new FirestoreFilterSpec
            {
                Collection = "procurement_requests",
                StatusIn = [ProcurementStatus.Approved, ProcurementStatus.Ordered],
                RequesterUidEquals = requesterUid,
                OrderBy = [new FirestoreOrderBy { Field = "updatedAtUtc", Descending = true }]
            },
            ProcurementTab.Rejected => new FirestoreFilterSpec
            {
                Collection = "procurement_requests",
                StatusIn = [ProcurementStatus.Rejected],
                RequesterUidEquals = requesterUid,
                OrderBy = [new FirestoreOrderBy { Field = "updatedAtUtc", Descending = true }]
            },
            ProcurementTab.HistoryReports => new FirestoreFilterSpec
            {
                Collection = "procurement_requests",
                StatusIn = [ProcurementStatus.Completed],
                RequesterUidEquals = requesterUid,
                OrderBy = [new FirestoreOrderBy { Field = "updatedAtUtc", Descending = true }]
            },
            ProcurementTab.StockMovements => FirestoreFilterSpec.ForStockMovements(readOnly: false),
            ProcurementTab.StockStatus => FirestoreFilterSpec.ForStockItems(readOnly: true),
            _ => null
        };
    }

    public static IReadOnlyList<string> StatusesForTab(ProcurementTab tab) => tab switch
    {
        ProcurementTab.PendingDraft => [ProcurementStatus.Draft, ProcurementStatus.Submitted],
        ProcurementTab.ManagementApproval => [ProcurementStatus.Submitted],
        ProcurementTab.QuoteRequested =>
        [
            ProcurementStatus.QuoteRequested,
            ProcurementStatus.QuoteEntry,
            ProcurementStatus.Comparison
        ],
        ProcurementTab.QuoteReview => [ProcurementStatus.ManagementQuoteReview],
        ProcurementTab.ApprovedOrders => [ProcurementStatus.Approved, ProcurementStatus.Ordered],
        ProcurementTab.Rejected => [ProcurementStatus.Rejected],
        ProcurementTab.HistoryReports => [ProcurementStatus.Completed],
        _ => []
    };

    public static bool MatchesTab(
        ProcurementTab tab,
        ProcurementRequestSnapshot request,
        string? role,
        string? currentUid)
    {
        if (!IsTabVisible(role, tab))
            return false;

        if (tab is ProcurementTab.StockMovements or ProcurementTab.StockStatus)
            return false;

        if (RequiresRequesterScope(role))
        {
            if (string.IsNullOrWhiteSpace(currentUid))
                return false;
            if (!request.RequesterUid.Equals(currentUid, StringComparison.OrdinalIgnoreCase))
                return false;
        }

        var normalized = request.NormalizedStatus;
        var allowed = StatusesForTab(tab);
        if (!allowed.Contains(normalized, StringComparer.OrdinalIgnoreCase))
            return false;

        return true;
    }

    public static IEnumerable<T> FilterForTab<T>(
        ProcurementTab tab,
        IEnumerable<T> source,
        Func<T, ProcurementRequestSnapshot> selector,
        string? role,
        string? currentUid)
    {
        if (!IsTabVisible(role, tab))
            return [];

        return source.Where(item => MatchesTab(tab, selector(item), role, currentUid));
    }

    public static IReadOnlyList<T> SortForTab<T>(
        ProcurementTab tab,
        IEnumerable<T> source,
        Func<T, ProcurementRequestSnapshot> selector)
    {
        var list = source.ToList();
        if (tab == ProcurementTab.ManagementApproval)
        {
            return list
                .OrderByDescending(item =>
                    selector(item).EffectivePriority.Equals(ProcurementPriority.Urgent, StringComparison.OrdinalIgnoreCase))
                .ThenByDescending(item => selector(item).Id)
                .ToList();
        }

        return list;
    }

    public static TabFilterContext CreateContext(string? role, string? currentUid) => new(role, currentUid);
}

public sealed class TabFilterContext
{
    public string RoleKey { get; }
    public string? CurrentUid { get; }
    public IReadOnlyList<ProcurementTab> VisibleTabs { get; }

    public TabFilterContext(string? role, string? currentUid)
    {
        RoleKey = TabFilterManager.NormalizeRole(role);
        CurrentUid = currentUid;
        VisibleTabs = TabFilterManager.GetVisibleTabs(role);
    }

    public bool IsVisible(ProcurementTab tab) => VisibleTabs.Contains(tab);

    public FirestoreFilterSpec? QueryFor(ProcurementTab tab) =>
        TabFilterManager.GetQuerySpec(tab, RoleKey, CurrentUid);

    public bool Matches(ProcurementTab tab, ProcurementRequestSnapshot request) =>
        TabFilterManager.MatchesTab(tab, request, RoleKey, CurrentUid);
}

File: SatinalmaPro.Android/app/src/main/java/com/satinalmapro/shared/filter/TabFilterManager.kt
package com.satinalmapro.shared.filter

import com.satinalmapro.android.core.roles.KullaniciRolleri
import com.satinalmapro.shared.model.PurchaseRequest

/**
 * Rol ve sekmeye göre Firestore sorgu şartlarını tek merkezden üretir.
 * Masaüstü (C# TabFilterManager) ile birebir aynı kurallar.
 */
object TabFilterManager {

    private val fullProcurementTabs = listOf(
        ProcurementTab.PENDING_DRAFT,
        ProcurementTab.MANAGEMENT_APPROVAL,
        ProcurementTab.QUOTE_REQUESTED,
        ProcurementTab.QUOTE_REVIEW,
        ProcurementTab.APPROVED_ORDERS,
        ProcurementTab.REJECTED,
        ProcurementTab.HISTORY_REPORTS
    )

    fun normalizeRole(role: String?): String {
        return when (KullaniciRolleri.normalize(role)) {
            KullaniciRolleri.ADMIN -> "admin"
            KullaniciRolleri.YONETIM -> "yonetim"
            KullaniciRolleri.SATINALMA -> "satinalma"
            KullaniciRolleri.SEF -> "sef"
            KullaniciRolleri.SAHA -> "saha"
            KullaniciRolleri.DEPO -> "depo"
            KullaniciRolleri.ATOLYE -> "atolye"
            else -> KullaniciRolleri.normalize(role).lowercase()
        }
    }

    /** Liste sorgularında oluşturan filtresi yok — firma içi tüm talepler görünür. */
    fun requiresRequesterScope(role: String?): Boolean = false

    fun getVisibleTabs(role: String?): List<ProcurementTab> {
        return when (normalizeRole(role)) {
            "depo" -> listOf(
                ProcurementTab.APPROVED_ORDERS,
                ProcurementTab.HISTORY_REPORTS,
                ProcurementTab.STOCK_MOVEMENTS
            )
            "atolye" -> listOf(ProcurementTab.STOCK_STATUS)
            "admin", "yonetim", "satinalma" -> fullProcurementTabs
            "sef", "saha" -> fullProcurementTabs
            else -> fullProcurementTabs
        }
    }

    fun isTabVisible(role: String?, tab: ProcurementTab): Boolean =
        getVisibleTabs(role).contains(tab)

    fun getQuerySpec(
        tab: ProcurementTab,
        role: String?,
        currentUid: String?
    ): FirestoreFilterSpec? {
        if (!isTabVisible(role, tab)) return null

        val requesterUid = if (requiresRequesterScope(role)) currentUid else null
        if (requiresRequesterScope(role) && requesterUid.isNullOrBlank()) return null

        return when (tab) {
            ProcurementTab.PENDING_DRAFT -> FirestoreFilterSpec(
                collection = "procurement_requests",
                statusIn = listOf(ProcurementStatus.DRAFT, ProcurementStatus.SUBMITTED),
                requesterUidEquals = requesterUid,
                orderBy = listOf(FirestoreOrderBy(field = "updatedAtUtc", descending = true))
            )
            ProcurementTab.MANAGEMENT_APPROVAL -> FirestoreFilterSpec(
                collection = "procurement_requests",
                statusIn = listOf(ProcurementStatus.SUBMITTED),
                requesterUidEquals = requesterUid,
                urgentFirst = true,
                orderBy = listOf(
                    FirestoreOrderBy(field = "priority", descending = true),
                    FirestoreOrderBy(field = "updatedAtUtc", descending = true)
                )
            )
            ProcurementTab.QUOTE_REQUESTED -> FirestoreFilterSpec(
                collection = "procurement_requests",
                statusIn = listOf(
                    ProcurementStatus.QUOTE_REQUESTED,
                    ProcurementStatus.QUOTE_ENTRY,
                    ProcurementStatus.COMPARISON
                ),
                requesterUidEquals = requesterUid,
                orderBy = listOf(FirestoreOrderBy(field = "updatedAtUtc", descending = true))
            )
            ProcurementTab.QUOTE_REVIEW -> FirestoreFilterSpec(
                collection = "procurement_requests",
                statusIn = listOf(ProcurementStatus.MANAGEMENT_QUOTE_REVIEW),
                requesterUidEquals = requesterUid,
                orderBy = listOf(FirestoreOrderBy(field = "updatedAtUtc", descending = true))
            )
            ProcurementTab.APPROVED_ORDERS -> FirestoreFilterSpec(
                collection = "procurement_requests",
                statusIn = listOf(ProcurementStatus.APPROVED, ProcurementStatus.ORDERED),
                requesterUidEquals = requesterUid,
                orderBy = listOf(FirestoreOrderBy(field = "updatedAtUtc", descending = true))
            )
            ProcurementTab.REJECTED -> FirestoreFilterSpec(
                collection = "procurement_requests",
                statusIn = listOf(ProcurementStatus.REJECTED),
                requesterUidEquals = requesterUid,
                orderBy = listOf(FirestoreOrderBy(field = "updatedAtUtc", descending = true))
            )
            ProcurementTab.HISTORY_REPORTS -> FirestoreFilterSpec(
                collection = "procurement_requests",
                statusIn = listOf(ProcurementStatus.COMPLETED),
                requesterUidEquals = requesterUid,
                orderBy = listOf(FirestoreOrderBy(field = "updatedAtUtc", descending = true))
            )
            ProcurementTab.STOCK_MOVEMENTS -> FirestoreFilterSpec.forStockMovements(readOnly = false)
            ProcurementTab.STOCK_STATUS -> FirestoreFilterSpec.forStockItems(readOnly = true)
        }
    }

    fun statusesForTab(tab: ProcurementTab): List<String> = when (tab) {
        ProcurementTab.PENDING_DRAFT -> listOf(ProcurementStatus.DRAFT, ProcurementStatus.SUBMITTED)
        ProcurementTab.MANAGEMENT_APPROVAL -> listOf(ProcurementStatus.SUBMITTED)
        ProcurementTab.QUOTE_REQUESTED -> listOf(
            ProcurementStatus.QUOTE_REQUESTED,
            ProcurementStatus.QUOTE_ENTRY,
            ProcurementStatus.COMPARISON
        )
        ProcurementTab.QUOTE_REVIEW -> listOf(ProcurementStatus.MANAGEMENT_QUOTE_REVIEW)
        ProcurementTab.APPROVED_ORDERS -> listOf(ProcurementStatus.APPROVED, ProcurementStatus.ORDERED)
        ProcurementTab.REJECTED -> listOf(ProcurementStatus.REJECTED)
        ProcurementTab.HISTORY_REPORTS -> listOf(ProcurementStatus.COMPLETED)
        ProcurementTab.STOCK_MOVEMENTS, ProcurementTab.STOCK_STATUS -> emptyList()
    }

    fun matchesTab(
        tab: ProcurementTab,
        request: ProcurementRequestSnapshot,
        role: String?,
        currentUid: String?
    ): Boolean {
        if (!isTabVisible(role, tab)) return false
        if (tab == ProcurementTab.STOCK_MOVEMENTS || tab == ProcurementTab.STOCK_STATUS) return false

        if (requiresRequesterScope(role)) {
            if (currentUid.isNullOrBlank()) return false
            if (!request.requesterUid.equals(currentUid, ignoreCase = true)) return false
        }

        val allowed = statusesForTab(tab)
        return allowed.any { it.equals(request.normalizedStatus, ignoreCase = true) }
    }

    fun <T> filterForTab(
        tab: ProcurementTab,
        source: List<T>,
        selector: (T) -> ProcurementRequestSnapshot,
        role: String?,
        currentUid: String?
    ): List<T> {
        if (!isTabVisible(role, tab)) return emptyList()
        return source.filter { item -> matchesTab(tab, selector(item), role, currentUid) }
    }

    fun <T> sortForTab(
        tab: ProcurementTab,
        source: List<T>,
        selector: (T) -> ProcurementRequestSnapshot
    ): List<T> {
        if (tab != ProcurementTab.MANAGEMENT_APPROVAL) return source
        return source.sortedWith(
            compareByDescending<T> {
                selector(it).effectivePriority.equals(ProcurementPriority.URGENT, ignoreCase = true)
            }.thenByDescending { selector(it).id }
        )
    }

    fun createContext(role: String?, currentUid: String?): TabFilterContext =
        TabFilterContext(role, currentUid)

    fun snapshotFrom(request: PurchaseRequest): ProcurementRequestSnapshot =
        ProcurementRequestSnapshot(
            id = request.id,
            status = request.status,
            requesterUid = request.requesterUid,
            priority = ProcurementPriority.fromRequestType(request.requestType),
            requestType = request.requestType
        )
}

class TabFilterContext(
    role: String?,
    val currentUid: String?
) {
    val roleKey: String = TabFilterManager.normalizeRole(role)
    val visibleTabs: List<ProcurementTab> = TabFilterManager.getVisibleTabs(role)

    fun isVisible(tab: ProcurementTab): Boolean = visibleTabs.contains(tab)

    fun queryFor(tab: ProcurementTab): FirestoreFilterSpec? =
        TabFilterManager.getQuerySpec(tab, roleKey, currentUid)

    fun matches(tab: ProcurementTab, request: ProcurementRequestSnapshot): Boolean =
        TabFilterManager.matchesTab(tab, request, roleKey, currentUid)
}
